src/queues/ThrottlingQueue.ts

```diff
@@ -0,0 +1,191 @@
+/*
+Copyright 2022 The Matrix.org Foundation C.I.C.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import { extractRequestError, LogLevel } from "matrix-bot-sdk";
+import { Mjolnir } from "../Mjolnir";
+
+export type Task<T> = (queue: ThrottlingQueue) => Promise<T>;
+
+/**
+ * A queue for backgrounding tasks without hammering servers too much.
+ */
+export class ThrottlingQueue {
+    /**
+     * The pending tasks.
+     */
+    private tasks: (() => Promise<void>)[] | null;
+
+    /**
+     * A timeout for the next task to execute.
+     */
+    private timeout: ReturnType<typeof setTimeout> | null;
+
+    /**
+     * How long we should wait between the completion of a tasks and the start of the next task.
+     * Any >=0 number is good.
+     */
+    private _delayMS: number;
+
+    /**
+     * Construct an empty queue.
+     *
+     * This queue will start executing whenever `push()` is called and stop
+     * whenever it runs out of tasks to execute.
+     *
+     * @param delayMS The default delay between executing two tasks, in ms.
+     */
+    constructor(private mjolnir: Mjolnir, delayMS: number) {
+        this.timeout = null;
+        this.delayMS = delayMS;
+        this.tasks = [];
+    }
+
+    /**
+     * Stop the queue, make sure we can never use it again.
+     */
+    public dispose() {
+        this.stop();
+        this.tasks = null;
+    }
+
+    /**
+     * The number of tasks waiting to be executed.
+     */
+    get length(): number {
+        return this.tasks!.length;
+    }
+
+    /**
+     * Push a new task onto the queue.
+     *
+     * @param task Some code to execute.
+     * @return A promise resolved/rejected once `task` is complete.
+     */
+    public push<T>(task: Task<T>): Promise<T> {
+        // Wrap `task` into a `Promise` to inform enqueuer when
+        // the task is complete.
+        return new Promise((resolve, reject) => {
+            const wrapper = async () => {
+                try {
+                    const result: T = await task(this);
+                    resolve(result);
+                } catch (ex) {
+                    reject(ex);
+                };
+            };
+            this.tasks!.push(wrapper);
+            this.start();
```

Comment:
Is the solution to https://github.com/matrix-org/mjolnir/pull/296/files#diff-55f1d3716255e750c792306912a3556bbf0d44e1af37ebd83bbe8ee4a8912170R35-R37 making sure these two final actions happen before the promise is resolved/rejected above (84&86)?

Reply:
I'm afraid I don't understand the question.

Reply:
I don't understand how the promises can be resolved too early in that test

Reply:
Which promises?

Reply:
Ah, right. Well, it's an integration test. Currently, it doesn't break, but it's meant to catch any error we could be introducing at a later stage.

Reply:
ohh, that makes sense